Not compatible with PMD 7.0.0

I upgraded to PMD 7.0.0 and got an error while building:
I am using gradle-cpd-plugin: 3.3.

`./gradlew cpdCheck -stacktrace`
Task :cpdCheck FAILED

 All completed @ Mon Mar 25 05:35:03 EDT 2024 

FAILURE: Build failed with an exception.

What went wrong:
Execution failed for task ':cpdCheck'.
 A failure occurred while executing de.aaschmid.gradle.plugins.cpd.internal.worker.CpdAction
    Could not create an instance of type de.aaschmid.gradle.plugins.cpd.internal.worker.CpdAction.
       Could not generate a decorated class for type CpdAction.
          net/sourceforge/pmd/cpd/Language

    Exception is:
    org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':cpdCheck'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:130)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:128)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:116)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
        at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:463)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
    Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing de.aaschmid.gradle.plugins.cpd.internal.worker.CpdAction
        at org.gradle.workers.internal.DefaultWorkerExecutor$WorkItemExecution.waitForCompletion(DefaultWorkerExecutor.java:287)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.lambda$waitForItemsAndGatherFailures$2(DefaultAsyncWorkTracker.java:130)
        at org.gradle.internal.Factories$1.create(Factories.java:31)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLocks(DefaultWorkerLeaseService.java:336)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLocks(DefaultWorkerLeaseService.java:319)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLock(DefaultWorkerLeaseService.java:324)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:126)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:92)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForAll(DefaultAsyncWorkTracker.java:78)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForCompletion(DefaultAsyncWorkTracker.java:66)
        at org.gradle.api.internal.tasks.execution.TaskExecution$3.run(TaskExecution.java:252)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.api.internal.tasks.execution.TaskExecution.executeAction(TaskExecution.java:229)
        at org.gradle.api.internal.tasks.execution.TaskExecution.executeActions(TaskExecution.java:212)
        at org.gradle.api.internal.tasks.execution.TaskExecution.executeWithPreviousOutputFiles(TaskExecution.java:195)
        at org.gradle.api.internal.tasks.execution.TaskExecution.execute(TaskExecution.java:162)
        at org.gradle.internal.execution.steps.ExecuteStep.executeInternal(ExecuteStep.java:105)
        at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:44)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:59)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:56)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:44)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:41)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:74)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:55)
        at org.gradle.internal.execution.steps.PreCreateOutputParentsStep.execute(PreCreateOutputParentsStep.java:50)
        at org.gradle.internal.execution.steps.PreCreateOutputParentsStep.execute(PreCreateOutputParentsStep.java:28)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:67)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:37)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:61)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:26)
        at org.gradle.internal.execution.steps.CaptureOutputsAfterExecutionStep.execute(CaptureOutputsAfterExecutionStep.java:67)
        at org.gradle.internal.execution.steps.CaptureOutputsAfterExecutionStep.execute(CaptureOutputsAfterExecutionStep.java:45)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:40)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:29)
        at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:189)
        at org.gradle.internal.execution.steps.BuildCacheStep.lambda$execute$1(BuildCacheStep.java:75)
        at org.gradle.internal.Either$Right.fold(Either.java:175)
        at org.gradle.internal.execution.caching.CachingState.fold(CachingState.java:62)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:73)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:48)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:46)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:35)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:76)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$2(SkipUpToDateStep.java:54)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:54)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
        at org.gradle.internal.execution.steps.ResolveIncrementalCachingStateStep.executeDelegate(ResolveIncrementalCachingStateStep.java:49)
        at org.gradle.internal.execution.steps.ResolveIncrementalCachingStateStep.executeDelegate(ResolveIncrementalCachingStateStep.java:27)
        at org.gradle.internal.execution.steps.AbstractResolveCachingStateStep.execute(AbstractResolveCachingStateStep.java:71)
        at org.gradle.internal.execution.steps.AbstractResolveCachingStateStep.execute(AbstractResolveCachingStateStep.java:39)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:65)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:36)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:106)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:55)
        at org.gradle.internal.execution.steps.AbstractCaptureStateBeforeExecutionStep.execute(AbstractCaptureStateBeforeExecutionStep.java:64)
        at org.gradle.internal.execution.steps.AbstractCaptureStateBeforeExecutionStep.execute(AbstractCaptureStateBeforeExecutionStep.java:43)
        at org.gradle.internal.execution.steps.AbstractSkipEmptyWorkStep.executeWithNonEmptySources(AbstractSkipEmptyWorkStep.java:125)
        at org.gradle.internal.execution.steps.AbstractSkipEmptyWorkStep.execute(AbstractSkipEmptyWorkStep.java:61)
        at org.gradle.internal.execution.steps.AbstractSkipEmptyWorkStep.execute(AbstractSkipEmptyWorkStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
        at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:36)
        at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:23)
        at org.gradle.internal.execution.steps.HandleStaleOutputsStep.execute(HandleStaleOutputsStep.java:75)
        at org.gradle.internal.execution.steps.HandleStaleOutputsStep.execute(HandleStaleOutputsStep.java:41)
        at org.gradle.internal.execution.steps.AssignMutableWorkspaceStep.lambda$execute$0(AssignMutableWorkspaceStep.java:35)
        at org.gradle.api.internal.tasks.execution.TaskExecution$4.withWorkspace(TaskExecution.java:289)
        at org.gradle.internal.execution.steps.AssignMutableWorkspaceStep.execute(AssignMutableWorkspaceStep.java:31)
        at org.gradle.internal.execution.steps.AssignMutableWorkspaceStep.execute(AssignMutableWorkspaceStep.java:22)
        at org.gradle.internal.execution.steps.ChoosePipelineStep.execute(ChoosePipelineStep.java:40)
        at org.gradle.internal.execution.steps.ChoosePipelineStep.execute(ChoosePipelineStep.java:23)
        at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.lambda$execute$2(ExecuteWorkBuildOperationFiringStep.java:67)
        at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.execute(ExecuteWorkBuildOperationFiringStep.java:67)
        at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.execute(ExecuteWorkBuildOperationFiringStep.java:39)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:46)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:34)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:48)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:35)
        at org.gradle.internal.execution.impl.DefaultExecutionEngine$1.execute(DefaultExecutionEngine.java:61)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:127)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:116)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
        at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:463)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
    Caused by: org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type de.aaschmid.gradle.plugins.cpd.internal.worker.CpdAction.
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:69)
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:55)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:61)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:54)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:48)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
        at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:48)
        at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
        at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
        at org.gradle.workers.internal.IsolatedClassloaderWorkerFactory$1.lambda$execute$0(IsolatedClassloaderWorkerFactory.java:57)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:44)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:41)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.workers.internal.AbstractWorker.executeWrappedInBuildOperation(AbstractWorker.java:41)
        at org.gradle.workers.internal.IsolatedClassloaderWorkerFactory$1.execute(IsolatedClassloaderWorkerFactory.java:49)
        at org.gradle.workers.internal.DefaultWorkerExecutor.lambda$submitWork$0(DefaultWorkerExecutor.java:174)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runExecution(DefaultConditionalExecutionQueue.java:187)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.access$700(DefaultConditionalExecutionQueue.java:120)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner$1.run(DefaultConditionalExecutionQueue.java:162)
        at org.gradle.internal.Factories$1.create(Factories.java:31)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:264)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:128)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:133)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runBatch(DefaultConditionalExecutionQueue.java:157)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.run(DefaultConditionalExecutionQueue.java:126)
        ... 2 more
    Caused by: org.gradle.internal.instantiation.ClassGenerationException: Could not generate a decorated class for type CpdAction.
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.generateUnderLock(AbstractClassGenerator.java:258)
        at org.gradle.cache.internal.DefaultCrossBuildInMemoryCacheFactory$AbstractCrossBuildInMemoryCache.get(DefaultCrossBuildInMemoryCacheFactory.java:130)
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.generate(AbstractClassGenerator.java:184)
        at org.gradle.internal.instantiation.generator.AsmBackedClassGenerator.generate(AsmBackedClassGenerator.java:121)
        at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.lambda$forType$0(Jsr330ConstructorSelector.java:56)
        at org.gradle.cache.Cache.lambda$get$0(Cache.java:31)
        at org.gradle.cache.internal.DefaultCrossBuildInMemoryCacheFactory$AbstractCrossBuildInMemoryCache.get(DefaultCrossBuildInMemoryCacheFactory.java:130)
        at org.gradle.cache.Cache.get(Cache.java:31)
        at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.forType(Jsr330ConstructorSelector.java:53)
        at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.forParams(Jsr330ConstructorSelector.java:48)
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:61)
        ... 33 more
    Caused by: java.lang.NoClassDefFoundError: net/sourceforge/pmd/cpd/Language
        at org.gradle.internal.reflect.ClassInspector.inspectClass(ClassInspector.java:70)
        at org.gradle.internal.reflect.ClassInspector.visitGraph(ClassInspector.java:56)
        at org.gradle.internal.reflect.ClassInspector.inspect(ClassInspector.java:36)
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.inspectType(AbstractClassGenerator.java:287)
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.generateUnderLock(AbstractClassGenerator.java:227)
        ... 43 more
    Caused by: java.lang.ClassNotFoundException: net.sourceforge.pmd.cpd.Language
        ... 48 more


Reviewing the PMD 7.0.0 release notes:  https://pmd.github.io/pmd/pmd_release_notes_pmd7.html, the class **net.sourceforge.pmd.cpd.Language** has been removed. 

```
The following classes have been removed:

pmd-core
net.sourceforge.pmd.cpd.AbstractLanguage
net.sourceforge.pmd.cpd.AnyLanguage
net.sourceforge.pmd.cpd.Language
net.sourceforge.pmd.cpd.LanguageFactory
net.sourceforge.pmd.cpd.MatchAlgorithm (now package private)
net.sourceforge.pmd.cpd.MatchCollector (now package private)
net.sourceforge.pmd.cpd.SourceCode (and all inner classes like FileCodeLoader, …)
net.sourceforge.pmd.cpd.token.TokenFilter
```

Examining the gradle-cpd-plugin code repository,  class  [**CpdAction.java**](https://github.com/aaschmid/gradle-cpd-plugin/blob/54d5fd312247b7a496ec7a713a2a76f59bde2c0a/src/main/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdAction.java) imports **net.sourceforge.pmd.cpd.Language** and other PMD/CPD classes that have been removed from PMD 7.0.0

```
package de.aaschmid.gradle.plugins.cpd.internal.worker;

import java.io.File;
import java.lang.reflect.UndeclaredThrowableException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;
import java.util.Properties;

import javax.inject.Inject;

import de.aaschmid.gradle.plugins.cpd.internal.worker.CpdWorkParameters.Report;
import net.sourceforge.pmd.cpd.AnyLanguage;
import net.sourceforge.pmd.cpd.CPDConfiguration;
import net.sourceforge.pmd.cpd.Language;
import net.sourceforge.pmd.cpd.LanguageFactory;
import net.sourceforge.pmd.cpd.Match;
import net.sourceforge.pmd.cpd.Tokenizer;
import org.gradle.api.GradleException;
import org.gradle.workers.WorkAction;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
```

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -2,7 +2,7 @@
 
 Contributions are very welcome. The following will provide some helpful guidelines.
 
-## JUnit dataprovider Contributor License Agreement
+## Gradle CPD plugin Contributor License Agreement
 
 * You will only submit contributions where you have authored 100% of the content.
 * You will only submit contributions to which you have the necessary rights.

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -14,7 +14,7 @@ from your employer to make the contributions.
 
 We love pull requests. Here is a quick guide:
 
-1. You need to have a JDK (at least version 1.8) installed.
+1. You need to have a JDK (at least version 11) installed.
 2. Fork the repo (see https://help.github.com/articles/fork-a-repo).
 3. Create a new branch from master.
 4. Ensure that you have a clean state by running `./gradlew clean build`.

--- a/README.md
+++ b/README.md
@@ -33,29 +33,29 @@ https://plugins.gradle.org/plugin/de.aaschmid.cpd.
 Requirements
 ------------
 
-Currently this plugin requires [PMD](https://pmd.github.io/) greater or equal to version 6.1.0 such that ```toolVersion >= '6.1.0'```.
+This plugin requires [PMD](https://pmd.github.io/) greater or equal to version 7.0.0 such that ```toolVersion >= '7.0.0'```.
 
-Explanation: v2.0 removes deprecated rendering API (= `net.sourceforge.pmd.cpd.Renderer`) and replace it with new
-`net.sourceforge.pmd.cpd.renderer.CPDRenderer` which was introduced with v6.1.0, see
-[PMD release notes](https://pmd.github.io/2018/02/25/PMD-6.1.0/#api-changes).
+Explanation: v7.0.0 changed the internal API, see [PMD release notes](https://pmd.github.io/pmd/pmd_release_notes_pmd7.html#-api).
+
+You can use older versions of this plugin if you want to use older versions of PMD (see table below).
 
 ### Supported versions
 
-| G. CPD plugin  | Gradle       | PMD         | Java¹  |
-|:-------------- |:------------ |:----------- |:------ |
-| [v0.1][]       | 1.10 - 4.x   | 5.0.0 - 5.x | 6 - 8  |
-| [v0.2][]       | 2.0 - 4.x    | 5.0.0 - 5.x | 6 - 8  |
-| [v0.4][]       | 2.3 - 4.x    | 5.2.0 - 5.x | 6 - 8  |
-| [v1.0][]       | 2.14 - 5.0   | 5.2.0 - 5.x | 6 - 8  |
-| [v1.1][]       | 2.14 - 5.0   | 5.2.2 - 6.x | 6 - 9  |
-| [v1.2][]       | >= 3.5.1     | >= 5.2.2    | >= 8   |
-| [v1.3][]       | >= 4.10.3    | >= 5.2.2    | >= 8   |
-| [v2.0][]       | 4.10.3 - 5.5 | >= 6.1.0    | >= 8   |
-| [v3.0][]       | 5.6 - 5.x    | >= 6.10.0   | >= 8   |
-| [v3.1][]       | >= 5.6       | >= 6.10.0   | >= 8   |
-| [v3.2][]       | >= 6.6       | >= 6.10.0   | >= 8   |
-| [v3.3][]       | >= 6.6       | >= 6.10.0   | >= 8   |
-| [v3.4][]       | >= 7.4       | >= 6.10.0   | >= 8   |
+| G. CPD plugin | Gradle       | PMD         | Java¹  |
+|:--------------|:------------ |:------------|:------ |
+| [v0.1][]      | 1.10 - 4.x   | 5.0.0 - 5.x | 6 - 8  |
+| [v0.2][]      | 2.0 - 4.x    | 5.0.0 - 5.x | 6 - 8  |
+| [v0.4][]      | 2.3 - 4.x    | 5.2.0 - 5.x | 6 - 8  |
+| [v1.0][]      | 2.14 - 5.0   | 5.2.0 - 5.x | 6 - 8  |
+| [v1.1][]      | 2.14 - 5.0   | 5.2.2 - 6.x | 6 - 9  |
+| [v1.2][]      | >= 3.5.1     | >= 5.2.2    | >= 8   |
+| [v1.3][]      | >= 4.10.3    | >= 5.2.2    | >= 8   |
+| [v2.0][]      | 4.10.3 - 5.5 | >= 6.1.0    | >= 8   |
+| [v3.0][]      | 5.6 - 5.x    | >= 6.10.0   | >= 8   |
+| [v3.1][]      | >= 5.6       | >= 6.10.0   | >= 8   |
+| [v3.2][]      | >= 6.6       | >= 6.10.0   | >= 8   |
+| [v3.3][]      | >= 6.6       | >= 6.10.0   | >= 8   |
+| [v3.4][]      | >= 7.4       | >= 7.0.0    | >= 8   |
 
 ¹: Java version may additionally depend on [PMD][]s version which is might not be properly reflected here.
 
@@ -67,7 +67,7 @@ This plugin is available using either the new [Gradle plugins DSL](https://gradl
 
 ```groovy
 plugins {
-    id 'de.aaschmid.cpd' version '3.3'
+    id 'de.aaschmid.cpd' version '3.4'
 }
 ```
 
@@ -77,27 +77,27 @@ from [Maven Central](http://search.maven.org/#search|ga|1|gradle-cpd-plugin) or
 ```groovy
 buildscript {
     repositories {
-        // choose your prefered one
+        // choose your preferred one
         mavenCentral()
         jcenter()
     }
 
     dependencies {
-        classpath 'de.aaschmid:gradle-cpd-plugin:3.3'
+        classpath 'de.aaschmid:gradle-cpd-plugin:3.4'
     }
 }
 apply plugin: 'de.aaschmid.cpd'
 ```
 
-**Attention:** The plugins groupId was changed from ```de.aaschmid.gradle.plugins``` to ```de.aaschmid``` in [v1.0][].
+**Attention:** The plugin's groupId was changed from ```de.aaschmid.gradle.plugins``` to ```de.aaschmid``` in [v1.0][].
 
-By default the copy-paste-detection looks at all source code of all projects which at least apply ```LifecycleBasePlugin```.
+By default, the copy-paste-detection looks at all source code of all projects which at least apply ```LifecycleBasePlugin```.
 If you use a different programming language and want to get it configured out of the box, please open an issue :-)
 
 ### Single module project
 
 If you have a single module project you just need to make sure that the ```LifecycleBasePlugin``` is also applied to it
-(explicitly or implicitly through e.g. Java or Groovy plugin). Otherwise you can simply add a dependency to a task you like by
+(explicitly or implicitly through e.g. Java or Groovy plugin). Otherwise, you can simply add a dependency to a task you like by
 
 ```groovy
 analyze.dependsOn(cpdCheck)
@@ -123,7 +123,7 @@ subprojects {
 
 ### Custom sourceSets
 
-If your are adding custom sourceSets (even in subProjects), it may occur that you either need to configure `cpdCheck`
+If you are adding custom sourceSets (even in subProjects), it may occur that you either need to configure `cpdCheck`
 afterwards or even manually configure `source`. Unfortunately, I have had problems with this case in the
 [CpdAcceptanceTest](https://github.com/aaschmid/gradle-cpd-plugin/blob/master/src/integTest/groovy/de/aaschmid/gradle/plugins/cpd/test/CpdAcceptanceTest.groovy)
 tests using Gradle [TestKit](https://docs.gradle.org/current/userguide/test_kit.html).
@@ -137,7 +137,7 @@ This example shows a project where only ```main``` sources should be checked for
 // optional - settings for every CPD task
 cpd {
     language = 'cpp'
-    toolVersion = '6.1.0' // defaults to '6.13.0'; just available for v6.1.0 and higher (see explanation above)
+    toolVersion = '7.0.0' // defaults to '7.2.0'; just available for v7.0.0 and higher (see explanation above)
 }
 
 // optional - default report is xml and default sources are 'main' and 'test'
@@ -150,7 +150,7 @@ cpdCheck {
 }
 ```
 
-*Note:* With v0.2, I have renamed the default task from ```cpd``` to ```cpdCheck``` that it does not have a name clash anymore.
+*Note:* With v0.2, I have renamed the default task from ```cpd``` to ```cpdCheck``` so that it does not have a name clash anymore.
 
 ### Property `source`
 
@@ -174,7 +174,7 @@ cpdCheck {
     }
 }
 ```
-or same using kotlin DSL (Note: That works only if all `subprojects` have `sourceSets` `main` and `test` e.g. by appling the `java` plugins. Otherwise you can just filter for `JavaBasePlugin` before)
+or same using kotlin DSL (Note: That works only if all `subprojects` have `sourceSets` `main` and `test` e.g. by appling the `java` plugins. Otherwise, you can just filter for `JavaBasePlugin` before)
 ```kotlin
 setSource(files(
     // only check java source code
@@ -229,8 +229,7 @@ e.g. using ```cpdCheck { }```:
 | skipBlocks         | ```true```           | ```'cpp'```                | [v0.4][] |
 | skipBlocksPattern  | ```'#if 0\|#endif'``` | ```'cpp'```               | [v0.4][] |
 
-If a specified `language` cannot be found, a fallback mechanism uses `net.sourceforge.pmd.cpd.AnyLanguage` instead. This
-fallback language does not run ANTLR and therefore also checks duplicates in comments.
+If a specified `language` cannot be found, analysis fails.
 
 For more information about options and their descriptions, see [here](https://pmd.github.io/latest/pmd_userdocs_cpd.html#attribute-reference),
 and for the available programming languages have a look on [CPD documentation](https://pmd.github.io/latest/pmd_userdocs_cpd.html#supported-languages).
@@ -278,3 +277,4 @@ get these dependencies in your ```localMaven()``` repository as the test cases l
 [v3.1]: /../../releases/tag/v3.1
 [v3.2]: /../../releases/tag/v3.2
 [v3.3]: /../../releases/tag/v3.3
+[v3.4]: /../../releases/tag/v3.4

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -30,9 +30,9 @@ sourceSets {
 }
 
 dependencies {
-    compileOnly("net.sourceforge.pmd:pmd-dist:6.10.0")
+    compileOnly("net.sourceforge.pmd:pmd-dist:7.2.0")
 
-    testImplementation("net.sourceforge.pmd:pmd-dist:6.10.0")
+    testImplementation("net.sourceforge.pmd:pmd-dist:7.2.0")
     testImplementation("com.google.guava:guava:33.1.0-jre")
     testImplementation("org.junit.jupiter:junit-jupiter:5.10.2")
     testImplementation("org.assertj:assertj-core:3.25.3")

--- a/src/integTest/groovy/de/aaschmid/gradle/plugins/cpd/test/CpdAcceptanceTest.groovy
+++ b/src/integTest/groovy/de/aaschmid/gradle/plugins/cpd/test/CpdAcceptanceTest.groovy
@@ -54,7 +54,7 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
         !file('build/reports/cpdCheck.csv').exists()
     }
 
-    def "Cpd will produce empty 'cpdCheck.xml' on non-duplicate 'java' source"() {
+    def "Cpd will produce 'cpdCheck.xml' without duplication information on non-duplicate 'java' source"() {
         given:
         buildFileWithPluginAndRepos() << """
             cpd{
@@ -74,7 +74,7 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
         def report = testProjectDir.resolve('build/reports/cpd/cpdCheck.xml').toFile()
         report.exists()
         report.text =~ /encoding="ISO-8859-1"/
-        report.text =~ /<pmd-cpd\/>/
+        !(report.text =~ /<duplication/)
     }
 
     @Issue("https://github.com/aaschmid/gradle-cpd-plugin/issues/38")
@@ -282,12 +282,12 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
         !csv.exists()
 
         txt.exists()
+        txt.text.contains("Found a 3 line (9 tokens) duplication in the following files:")
         txt.text.contains("Found a 4 line (9 tokens) duplication in the following files:")
-        txt.text.contains("Found a 2 line (8 tokens) duplication in the following files:")
 
         xml.exists()
+        xml.text.contains('<duplication lines="3" tokens="9">')
         xml.text.contains('<duplication lines="4" tokens="9">')
-        xml.text.contains('<duplication lines="2" tokens="8">')
     }
 
     @Issue("https://github.com/aaschmid/gradle-cpd-plugin/issues/39")
@@ -337,8 +337,8 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
 
         def kotlinReport = file('kotlin-cpd.xml')
         kotlinReport.exists()
+        kotlinReport.text.contains('<duplication lines="3" tokens="9">')
         kotlinReport.text.contains('<duplication lines="4" tokens="9">')
-        kotlinReport.text.contains('<duplication lines="2" tokens="8">')
     }
 
     def "Cpd should fail if not ignoreAnnotations on duplicate annotations"() {
@@ -366,7 +366,7 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
         def report = file('build/reports/cpd/cpdCheck.csv')
         report.exists()
         // locally Person.java comes before Employee, on travis-ci is Employee first => make it irrelevant
-        report.text =~ /8,53,2,6,.*(Person|Employee)\.java,6,.*(Person|Employee)\.java/
+        report.text =~ /8,54,2,6,.*(Person|Employee)\.java,6,.*(Person|Employee)\.java/
         report.text =~ /14,45,2,13,.*(Person|Employee)\.java,13,.*(Person|Employee)\.java/
     }
 
@@ -443,7 +443,7 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
 
         def report = file('build/reports/cpd/cpdCheck.xml')
         report.exists()
-        report.text =~ /<pmd-cpd\/>/
+        !(report.text =~ /<duplication/)
     }
 
     def "Cpd should fail if ignoreLiterals on different literals"() {
@@ -469,11 +469,11 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
 
         def vsReport = file('build/reports/cpd/cpdCheck.vs')
         vsReport.exists()
-        vsReport.text =~ /Literal[12].java\(5\): Between lines 5 and 14/
+        vsReport.text =~ /Literal[12].java\(5\): Between lines 5 and 13/
 
         def xmlReport = file('build/reports/cpd/cpdCheck.xml')
         xmlReport.exists()
-        xmlReport.text =~ /<duplication lines="9" tokens="27">\s+<file line="5"\s+path=".*Literal[12].java"\/>/
+        xmlReport.text =~ /<duplication lines="9" tokens="27">\s+<file[^>]*line="5"\s+path=".*Literal[12].java"\/>/
     }
 
     def "Cpd should not fail if not ignoreLiterals on different literals"() {
@@ -495,7 +495,7 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
 
         def report = file('build/reports/cpd/cpdCheck.xml')
         report.exists()
-        report.text =~ /<pmd-cpd\/>/
+        !(report.text =~ /<duplication/)
     }
 
     def "Cpd should fail if not skipDuplicateFiles on duplicate files"() {
@@ -544,7 +544,7 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
 
         def report = file('build/reports/cpd/cpdCheck.xml')
         report.exists()
-        report.text =~ /<pmd-cpd\/>/
+        !(report.text =~ /<duplication/)
     }
 
 
@@ -563,7 +563,7 @@ class CpdAcceptanceTest extends IntegrationBaseSpec {
         then:
         result.task(':cpdCheck').outcome == FAILED
         result.output.contains("BUILD FAILED")
-        result.output =~ /Lexical error in file .*Error.java at/
+        result.output =~ /Lexical error in file '.*Error.java' at/
 
         def report = file('build/reports/cpd/cpdCheck.csv')
         !report.exists()

--- a/src/integTest/groovy/de/aaschmid/gradle/plugins/cpd/test/CpdIntegrationTest.groovy
+++ b/src/integTest/groovy/de/aaschmid/gradle/plugins/cpd/test/CpdIntegrationTest.groovy
@@ -182,15 +182,15 @@ class CpdIntegrationTest extends IntegrationBaseSpec {
 
         def report = file('build/reports/cpd/cpdCheck.xml')
         report.exists()
+        report.text.contains('<duplication lines="3" tokens="9">')
         report.text.contains('<duplication lines="4" tokens="9">')
-        report.text.contains('<duplication lines="2" tokens="8">')
     }
 
     def "executing 'Cpd' task on duplicate 'java' source with minimal supported PMD version should produce 'cpdCheck.xml'"() {
         given:
         buildFileWithPluginAndRepos() << """
             cpd{
-                toolVersion = '6.10.0'
+                toolVersion = '7.0.0'
             }
             cpdCheck{
                 ignoreFailures = true
@@ -215,16 +215,13 @@ class CpdIntegrationTest extends IntegrationBaseSpec {
         report.text =~ /4,15,2,[79],.*Clazz[12]\.java,[79],.*Clazz[12]\.java/
     }
 
-    def "executing 'Cpd' task on duplicates should fallback to any language if language does not exist"() {
+    def "executing 'Cpd' task on duplicates should fail if language does not exist"() {
         given:
         buildFile << createBuildScriptWithClasspathOfGradleTestKitMechanism() << """
             apply plugin: 'de.aaschmid.cpd'
             repositories {
                 mavenLocal()
                 mavenCentral()
-            }
-            cpd{
-                toolVersion = '6.1.0'
             }
             cpdCheck{
                 language = 'my-lang'
@@ -237,11 +234,6 @@ class CpdIntegrationTest extends IntegrationBaseSpec {
 
         then:
         result.output.contains("BUILD FAILED")
-        result.output.contains("Could not detect CPD language for 'my-lang', using 'any' as fallback language.")
-        result.output =~ /CPD found duplicate code. See the report at file:\/.*\/cpdCheck.xml/
-
-        def report = file('build/reports/cpd/cpdCheck.xml')
-        report.exists()
-        report.text.contains('<duplication lines="19" tokens="118">')
+        result.output.contains("Could not detect CPD language for 'my-lang'.")
     }
 }

--- a/src/main/java/de/aaschmid/gradle/plugins/cpd/CpdExtension.java
+++ b/src/main/java/de/aaschmid/gradle/plugins/cpd/CpdExtension.java
@@ -183,7 +183,7 @@ public class CpdExtension extends CodeQualityExtension {
     /**
      * CConfigures the pattern, to find the blocks to skip if enabled using {@link #skipBlocks}. It is a {@link String} property and
      * contains of two parts, separated by {@code '|'}. The first part is the start pattern, the second part is the ending pattern. defaults
-     * to {@code '#if 0|#endif'} (which should be the same as {@link net.sourceforge.pmd.cpd.Tokenizer#DEFAULT_SKIP_BLOCKS_PATTERN}).
+     * to {@code '#if 0|#endif'} (which should be the same as {@link net.sourceforge.pmd.cpd.internal.CpdLanguagePropertiesDefaults#DEFAULT_SKIP_BLOCKS_PATTERN}).
      * <p>
      * Example: {@code skipBlocksPattern = '#include <|>'}
      *

--- a/src/main/java/de/aaschmid/gradle/plugins/cpd/CpdPlugin.java
+++ b/src/main/java/de/aaschmid/gradle/plugins/cpd/CpdPlugin.java
@@ -16,6 +16,7 @@ import org.gradle.api.logging.Logger;
 import org.gradle.api.logging.Logging;
 import org.gradle.api.plugins.JavaBasePlugin;
 import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.plugins.JvmEcosystemPlugin;
 import org.gradle.api.plugins.ReportingBasePlugin;
 import org.gradle.api.provider.ProviderFactory;
 import org.gradle.api.reporting.ReportingExtension;
@@ -74,6 +75,7 @@ public class CpdPlugin implements Plugin<Project> {
     @Override
     public void apply(Project project) {
         project.getPlugins().apply(ReportingBasePlugin.class);
+        project.getPlugins().apply(JvmEcosystemPlugin.class);
 
         CpdExtension extension = createExtension(project);
         createConfiguration(project, extension);
@@ -86,7 +88,7 @@ public class CpdPlugin implements Plugin<Project> {
 
     private CpdExtension createExtension(Project project) {
         CpdExtension extension = project.getExtensions().create("cpd", CpdExtension.class);
-        extension.setToolVersion("6.14.0");
+        extension.setToolVersion("7.2.0");
         return extension;
     }
 

--- a/src/main/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdAction.java
+++ b/src/main/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdAction.java
@@ -4,18 +4,17 @@ import java.io.File;
 import java.lang.reflect.UndeclaredThrowableException;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.nio.charset.Charset;
 import java.util.List;
-import java.util.Properties;
 
 import javax.inject.Inject;
 
 import de.aaschmid.gradle.plugins.cpd.internal.worker.CpdWorkParameters.Report;
-import net.sourceforge.pmd.cpd.AnyLanguage;
 import net.sourceforge.pmd.cpd.CPDConfiguration;
-import net.sourceforge.pmd.cpd.Language;
-import net.sourceforge.pmd.cpd.LanguageFactory;
+import net.sourceforge.pmd.cpd.CPDReport;
+import net.sourceforge.pmd.lang.Language;
 import net.sourceforge.pmd.cpd.Match;
-import net.sourceforge.pmd.cpd.Tokenizer;
+import net.sourceforge.pmd.lang.LanguageRegistry;
 import org.gradle.api.GradleException;
 import org.gradle.workers.WorkAction;
 import org.slf4j.Logger;
@@ -42,15 +41,20 @@ public abstract class CpdAction implements WorkAction<CpdWorkParameters> {
 
     @Override
     public void execute() {
-        List<Match> matches = executor.run(createCpdConfiguration(getParameters()), getParameters().getSourceFiles().getFiles());
-        reporter.generate(getParameters().getReportParameters().get(), matches);
-        logResult(matches);
+        CPDReport cpdReport = executor.run(createCpdConfiguration(getParameters()), getParameters().getSourceFiles().getFiles());
+        reporter.generate(getParameters().getReportParameters().get(), cpdReport);
+        logResult(cpdReport.getMatches());
     }
 
     private CPDConfiguration createCpdConfiguration(CpdWorkParameters config) {
         CPDConfiguration result = new CPDConfiguration();
-        result.setEncoding(config.getEncoding().get());
-        result.setLanguage(createLanguage(config.getLanguage().get(), createLanguageProperties(config)));
+        result.setSourceEncoding(getEncoding(config));
+        result.setIgnoreAnnotations(config.getIgnoreAnnotations().get());
+        result.setIgnoreIdentifiers(config.getIgnoreIdentifiers().get());
+        result.setIgnoreLiterals(config.getIgnoreLiterals().get());
+        result.setNoSkipBlocks(!config.getSkipBlocks().get());
+        result.setSkipBlocksPattern(config.getSkipBlocksPattern().get());
+        result.setOnlyRecognizeLanguage(createLanguage(config.getLanguage().get()));
         result.setMinimumTileSize(config.getMinimumTokenCount().get());
         result.setSkipDuplicates(config.getSkipDuplicateFiles().get());
         result.setSkipLexicalErrors(config.getSkipLexicalErrors().get());
@@ -79,40 +83,28 @@ public abstract class CpdAction implements WorkAction<CpdWorkParameters> {
         }
     }
 
-    private Properties createLanguageProperties(CpdWorkParameters config) {
-        Properties languageProperties = new Properties();
-
-        if (config.getIgnoreAnnotations().get()) {
-            languageProperties.setProperty(Tokenizer.IGNORE_ANNOTATIONS, "true");
+    private Language createLanguage(String language) {
+        Language result = LanguageRegistry.CPD.getLanguageById(language);
+        if (result == null) {
+            throw new GradleException(String.format("Could not detect CPD language for '%s'.", language));
         }
-        if (config.getIgnoreIdentifiers().get()) {
-            languageProperties.setProperty(Tokenizer.IGNORE_IDENTIFIERS, "true");
-        }
-        if (config.getIgnoreLiterals().get()) {
-            languageProperties.setProperty(Tokenizer.IGNORE_LITERALS, "true");
-        }
-        languageProperties.setProperty(Tokenizer.OPTION_SKIP_BLOCKS, Boolean.toString(config.getSkipBlocks().get()));
-        languageProperties.setProperty(Tokenizer.OPTION_SKIP_BLOCKS_PATTERN, config.getSkipBlocksPattern().get());
-        return languageProperties;
-    }
-
-    private Language createLanguage(String language, Properties languageProperties) {
-        Language result = LanguageFactory.createLanguage(language, languageProperties);
         logger.info("Using CPD language class '{}' for checking duplicates.", result);
-        if (result instanceof AnyLanguage) {
-            logger.warn("Could not detect CPD language for '{}', using 'any' as fallback language.", language);
-        }
         return result;
     }
 
     private String asClickableFileUrl(File path) {
         // File.toURI().toString() leads to an URL like this on Mac: file:/reports/index.html
         // This URL is not recognized by the Mac console (too few leading slashes). We solve
-        // this be creating an URI with an empty authority.
+        // this be creating a URI with an empty authority.
         try {
             return new URI("file", "", path.toURI().getPath(), null, null).toString();
         } catch (URISyntaxException e) {
             throw new UndeclaredThrowableException(e);
         }
+    }
+
+    private Charset getEncoding(CpdWorkParameters config) {
+        String encodingString = config.getEncoding().get();
+        return Charset.forName(encodingString);
     }
 }

--- a/src/main/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdExecutor.java
+++ b/src/main/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdExecutor.java
@@ -2,34 +2,31 @@ package de.aaschmid.gradle.plugins.cpd.internal.worker;
 
 import java.io.File;
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
 import java.util.Set;
-import java.util.Spliterator;
 
-import net.sourceforge.pmd.cpd.CPD;
 import net.sourceforge.pmd.cpd.CPDConfiguration;
-import net.sourceforge.pmd.cpd.Match;
+import net.sourceforge.pmd.cpd.CPDReport;
+import net.sourceforge.pmd.cpd.CpdAnalysis;
 import org.gradle.api.GradleException;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.logging.Logging;
-
-import static java.util.Spliterators.spliteratorUnknownSize;
-import static java.util.stream.Collectors.toList;
-import static java.util.stream.StreamSupport.stream;
 
 class CpdExecutor {
 
     private static final Logger logger = Logging.getLogger(CpdExecutor.class);
 
-    List<Match> run(CPDConfiguration cpdConfig, Set<File> sourceFiles) {
+    CPDReport run(CPDConfiguration cpdConfig, Set<File> sourceFiles) {
         if (logger.isInfoEnabled()) {
             logger.info("Starting CPD, minimumTokenCount is {}", cpdConfig.getMinimumTileSize());
         }
-        try {
-            CPD cpd = new CPD(cpdConfig);
+
+        try (CpdAnalysis cpd = CpdAnalysis.create(cpdConfig)) {
             tokenizeSourceFiles(cpd, sourceFiles);
-            analyzeSourceCode(cpd);
-            return stream(spliteratorUnknownSize(cpd.getMatches(), Spliterator.ORDERED), false).collect(toList());
+            return analyzeSourceCode(cpd)
+                    .orElseThrow(() -> new GradleException("Analysis did not produce any result"));
         } catch (IOException e) {
             throw new GradleException("Exception during CPD execution: " + e.getMessage(), e);
         } catch (Throwable t) {
@@ -37,26 +34,31 @@ class CpdExecutor {
         }
     }
 
-    private void tokenizeSourceFiles(CPD cpd, Set<File> sourceFiles) throws IOException {
+    private void tokenizeSourceFiles(CpdAnalysis cpd, Set<File> sourceFiles) throws IOException {
         for (File file : sourceFiles) {
             if (logger.isDebugEnabled()) {
                 logger.debug("Tokenize {}", file.getAbsolutePath());
             }
-            cpd.add(file);
+            cpd.files().addFile(file.toPath());
         }
     }
 
-    private void analyzeSourceCode(CPD cpd) {
+    private Optional<CPDReport> analyzeSourceCode(CpdAnalysis cpd) {
         if (logger.isInfoEnabled()) {
             logger.info("Starting to analyze code");
         }
+        List<CPDReport> cpdReports = new ArrayList<>();
         long start = System.currentTimeMillis();
-        cpd.go();
+        cpd.performAnalysis(cpdReports::add);
         long stop = System.currentTimeMillis();
 
         long timeTaken = stop - start;
         if (logger.isInfoEnabled()) {
             logger.info("Successfully analyzed code - took {} milliseconds", timeTaken);
         }
+        if (cpdReports.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(cpdReports.get(0));
     }
 }

--- a/src/main/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdReporter.java
+++ b/src/main/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdReporter.java
@@ -7,12 +7,12 @@ import java.util.List;
 
 import de.aaschmid.gradle.plugins.cpd.internal.worker.CpdWorkParameters.Report;
 import de.aaschmid.gradle.plugins.cpd.internal.worker.CpdWorkParameters.Report.Xml;
+import net.sourceforge.pmd.cpd.CPDReport;
+import net.sourceforge.pmd.cpd.CPDReportRenderer;
 import net.sourceforge.pmd.cpd.CSVRenderer;
-import net.sourceforge.pmd.cpd.Match;
 import net.sourceforge.pmd.cpd.SimpleRenderer;
 import net.sourceforge.pmd.cpd.VSRenderer;
 import net.sourceforge.pmd.cpd.XMLRenderer;
-import net.sourceforge.pmd.cpd.renderer.CPDRenderer;
 import org.gradle.api.GradleException;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.logging.Logging;
@@ -21,12 +21,12 @@ class CpdReporter {
 
     private static final Logger logger = Logging.getLogger(CpdReporter.class);
 
-    void generate(List<Report> reports, List<Match> matches) {
+    void generate(List<Report> reports, CPDReport cpdReport) {
         if (logger.isInfoEnabled()) {
             logger.info("Generating reports");
         }
         for (Report report : reports) {
-            CPDRenderer renderer = createRendererFor(report);
+            CPDReportRenderer renderer = createRendererFor(report);
             try (FileWriter fileWriter = new FileWriter(report.getDestination())) {
 
                 ClassLoader previousContextClassLoader = Thread.currentThread().getContextClassLoader();
@@ -34,7 +34,7 @@ class CpdReporter {
                     // Workaround for Gradle Worker API using special class loader which Xerces dynamic implementation loading does not like
                     Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
 
-                    renderer.render(matches.iterator(), fileWriter);
+                    renderer.render(cpdReport, fileWriter);
                 } finally {
                     Thread.currentThread().setContextClassLoader(previousContextClassLoader);
                 }
@@ -46,12 +46,12 @@ class CpdReporter {
 
     /**
      * Note: This cannot be implemented in {@link Report} subclasses because they must be independent of {@link
-     * CPDRenderer} because only worker classloader knows about PMD / CPD library.
+     * CPDReportRenderer} because only worker classloader knows about PMD / CPD library.
      *
      * @param report the configured reports used
-     * @return a full configured {@link CPDRenderer} to generate a CPD single file reports.
+     * @return a full configured {@link CPDReportRenderer} to generate a CPD single file reports.
      */
-    CPDRenderer createRendererFor(Report report) {
+    CPDReportRenderer createRendererFor(Report report) {
         if (report instanceof Report.Csv) {
             char separator = ((Report.Csv) report).getSeparator();
             boolean lineCountPerFile = !((Report.Csv) report).isIncludeLineCount();
@@ -92,7 +92,7 @@ class CpdReporter {
      * <i>Information:</i> Use reflection because neither proper constructor for setting both fields nor setter are
      * available.
      */
-    private void setTrimLeadingWhitespacesByReflection(CPDRenderer result, boolean trimLeadingCommonSourceWhitespaces) {
+    private void setTrimLeadingWhitespacesByReflection(CPDReportRenderer result, boolean trimLeadingCommonSourceWhitespaces) {
         String fieldName = "trimLeadingWhitespace";
         if (logger.isDebugEnabled()) {
             logger.debug("Try setting '{}' field to '{}' for '{}' by reflection.", fieldName, trimLeadingCommonSourceWhitespaces, result);

--- a/src/test/java/de/aaschmid/gradle/plugins/cpd/CpdPluginTest.java
+++ b/src/test/java/de/aaschmid/gradle/plugins/cpd/CpdPluginTest.java
@@ -7,7 +7,7 @@ import java.util.Set;
 import java.util.stream.Stream;
 
 import de.aaschmid.gradle.plugins.cpd.test.GradleExtension;
-import net.sourceforge.pmd.cpd.Tokenizer;
+import net.sourceforge.pmd.cpd.internal.CpdLanguagePropertiesDefaults;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
@@ -56,8 +56,8 @@ class CpdPluginTest {
         assertThat(cpd.isSkipDuplicateFiles()).isFalse();
         assertThat(cpd.isSkipLexicalErrors()).isFalse();
         assertThat(cpd.isSkipBlocks()).isTrue();
-        assertThat(cpd.getSkipBlocksPattern()).isEqualTo(Tokenizer.DEFAULT_SKIP_BLOCKS_PATTERN);
-        assertThat(cpd.getToolVersion()).isEqualTo("6.14.0");
+        assertThat(cpd.getSkipBlocksPattern()).isEqualTo(CpdLanguagePropertiesDefaults.DEFAULT_SKIP_BLOCKS_PATTERN);
+        assertThat(cpd.getToolVersion()).isEqualTo("7.2.0");
     }
 
     @Test
@@ -102,7 +102,7 @@ class CpdPluginTest {
         assertThat(t.getSkipDuplicateFiles()).isFalse();
         assertThat(t.getSkipLexicalErrors()).isFalse();
         assertThat(t.getSkipBlocks()).isTrue();
-        assertThat(t.getSkipBlocksPattern()).isEqualTo(Tokenizer.DEFAULT_SKIP_BLOCKS_PATTERN);
+        assertThat(t.getSkipBlocksPattern()).isEqualTo(CpdLanguagePropertiesDefaults.DEFAULT_SKIP_BLOCKS_PATTERN);
 
         assertThat(t.getSource()).isEmpty();
     }
@@ -139,11 +139,12 @@ class CpdPluginTest {
         assertThat(t.getSkipDuplicateFiles()).isFalse();
         assertThat(t.getSkipLexicalErrors()).isFalse();
         assertThat(t.getSkipBlocks()).isTrue();
-        assertThat(t.getSkipBlocksPattern()).isEqualTo(Tokenizer.DEFAULT_SKIP_BLOCKS_PATTERN);
+        assertThat(t.getSkipBlocksPattern()).isEqualTo(CpdLanguagePropertiesDefaults.DEFAULT_SKIP_BLOCKS_PATTERN);
 
         assertThat(t.getSource()).isEmpty();
     }
 
+    @SuppressWarnings("UnstableApiUsage")
     static Stream<Class<? extends Plugin<?>>> CpdPlugin_shouldAddCpdCheckTaskAsDependencyOfCheckLifecycleTaskIfPluginIsApplied() {
         return Stream.of(
                 LifecycleBasePlugin.class,
@@ -182,8 +183,10 @@ class CpdPluginTest {
         Task checkTask = project.getTasks().getByName("check");
 
         // Then:
-        assertThat((Set<Task>) checkTask.getTaskDependencies().getDependencies(checkTask)).contains(cpdCheck.get());
-        assertThat(cpdCheck.get().getSource()).containsExactlyInAnyOrderElementsOf(testFilesRecurseIn(JAVA, "."));
+        Cpd cpdTask = cpdCheck.get();
+        //noinspection unchecked
+        assertThat((Set<Task>) checkTask.getTaskDependencies().getDependencies(checkTask)).contains(cpdTask);
+        assertThat(cpdTask.getSource()).containsExactlyInAnyOrderElementsOf(testFilesRecurseIn(JAVA, "."));
     }
 
     @Test

--- a/src/test/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdActionTest.java
+++ b/src/test/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdActionTest.java
@@ -1,18 +1,11 @@
 package de.aaschmid.gradle.plugins.cpd.internal.worker;
 
-import java.io.File;
-import java.util.List;
-import java.util.Set;
-
 import de.aaschmid.gradle.plugins.cpd.internal.worker.CpdWorkParameters.Report;
 import de.aaschmid.gradle.plugins.cpd.test.GradleExtension;
 import net.sourceforge.pmd.cpd.CPDConfiguration;
-import net.sourceforge.pmd.cpd.CPPLanguage;
-import net.sourceforge.pmd.cpd.CPPTokenizer;
-import net.sourceforge.pmd.cpd.JavaLanguage;
-import net.sourceforge.pmd.cpd.JavaTokenizer;
-import net.sourceforge.pmd.cpd.KotlinLanguage;
+import net.sourceforge.pmd.cpd.CPDReport;
 import net.sourceforge.pmd.cpd.Match;
+import net.sourceforge.pmd.lang.LanguageVersionDiscoverer;
 import org.gradle.api.GradleException;
 import org.gradle.api.Project;
 import org.junit.jupiter.api.BeforeEach;
@@ -24,6 +17,11 @@ import org.mockito.InOrder;
 import org.mockito.Mock;
 import org.mockito.Mockito;
 import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Set;
 
 import static de.aaschmid.gradle.plugins.cpd.test.PropertyUtils.listProperty;
 import static de.aaschmid.gradle.plugins.cpd.test.PropertyUtils.property;
@@ -63,6 +61,7 @@ class CpdActionTest {
                 return parameters;
             }
         };
+        when(executor.run(any(), any())).thenReturn(mock(CPDReport.class));
     }
 
     @Test
@@ -71,8 +70,8 @@ class CpdActionTest {
         Set<File> sourceFiles = singleton(testFile(JAVA, "de/aaschmid/clazz/Clazz.java"));
         List<Report> reports = singletonList(new Report.Csv(new File("cpd.csv"), ';', true));
 
-        List<Match> matches = singletonList(mock(Match.class));
-        when(executor.run(any(), eq(sourceFiles))).thenReturn(matches);
+        CPDReport cpdReport = mockReportFor(mock(Match.class));
+        when(executor.run(any(), eq(sourceFiles))).thenReturn(cpdReport);
 
         stubParametersWithDefaults(project);
         when(parameters.getIgnoreFailures()).thenReturn(property(true));
@@ -85,7 +84,7 @@ class CpdActionTest {
         // Then:
         InOrder inOrder = Mockito.inOrder(executor, reporter);
         inOrder.verify(executor).run(cpdConfiguration.capture(), eq(sourceFiles));
-        inOrder.verify(reporter).generate(reports, matches);
+        inOrder.verify(reporter).generate(reports, cpdReport);
         inOrder.verifyNoMoreInteractions();
     }
 
@@ -94,8 +93,8 @@ class CpdActionTest {
         // Given:
         Report.Xml report = new Report.Xml(new File("cpd.xml"), "UTF-8");
 
-        List<Match> matches = singletonList(mock(Match.class));
-        when(executor.run(any(), any())).thenReturn(matches);
+        CPDReport cpdReport = mockReportFor(mock(Match.class));
+        when(executor.run(any(), any())).thenReturn(cpdReport);
 
         stubParametersWithDefaults(project);
         when(parameters.getIgnoreFailures()).thenReturn(property(false));
@@ -124,8 +123,8 @@ class CpdActionTest {
         verify(executor).run(cpdConfiguration.capture(), any());
 
         CPDConfiguration actualCpdConfig = cpdConfiguration.getValue();
-        assertThat(actualCpdConfig.getEncoding()).isEqualTo("US-ASCII");
-        assertThat(actualCpdConfig.getLanguage()).isInstanceOf(KotlinLanguage.class);
+        assertThat(actualCpdConfig.getSourceEncoding()).isEqualTo(StandardCharsets.US_ASCII);
+        assertLanguage(actualCpdConfig, "kotlin");
         assertThat(actualCpdConfig.getMinimumTileSize()).isEqualTo(15);
         assertThat(actualCpdConfig.isSkipDuplicates()).isTrue();
         assertThat(actualCpdConfig.isSkipLexicalErrors()).isTrue();
@@ -147,12 +146,10 @@ class CpdActionTest {
         verify(executor).run(cpdConfiguration.capture(), any());
 
         CPDConfiguration actualCpdConfig = cpdConfiguration.getValue();
-        assertThat(actualCpdConfig.getLanguage()).isInstanceOf(JavaLanguage.class);
-        assertThat(actualCpdConfig.getLanguage().getTokenizer())
-                .isInstanceOf(JavaTokenizer.class)
-                .hasFieldOrPropertyWithValue("ignoreAnnotations", true)
-                .hasFieldOrPropertyWithValue("ignoreIdentifiers", true)
-                .hasFieldOrPropertyWithValue("ignoreLiterals", true);
+        assertLanguage(actualCpdConfig, "java");
+        assertThat(actualCpdConfig.isIgnoreAnnotations()).isTrue();
+        assertThat(actualCpdConfig.isIgnoreIdentifiers()).isTrue();
+        assertThat(actualCpdConfig.isIgnoreLiterals()).isTrue();
     }
 
     @Test
@@ -170,12 +167,9 @@ class CpdActionTest {
         verify(executor).run(cpdConfiguration.capture(), any());
 
         CPDConfiguration actualCpdConfig = cpdConfiguration.getValue();
-        assertThat(actualCpdConfig.getLanguage()).isInstanceOf(CPPLanguage.class);
-        assertThat(actualCpdConfig.getLanguage().getTokenizer())
-                .isInstanceOf(CPPTokenizer.class)
-                .hasFieldOrPropertyWithValue("skipBlocks", true)
-                .hasFieldOrPropertyWithValue("skipBlocksStart", "template<")
-                .hasFieldOrPropertyWithValue("skipBlocksEnd", ">");
+        assertLanguage(actualCpdConfig, "cpp");
+        assertThat(actualCpdConfig.isNoSkipBlocks()).isFalse();
+        assertThat(actualCpdConfig.getSkipBlocksPattern()).isEqualTo("template<|>");
     }
 
     private void stubParametersWithDefaults(Project project) {
@@ -194,5 +188,33 @@ class CpdActionTest {
         when(parameters.getSkipLexicalErrors()).thenReturn(property(false));
         when(parameters.getSourceFiles()).thenReturn(project.files(sourceFiles));
         when(parameters.getReportParameters()).thenReturn(listProperty(Report.class, singletonList(report)));
+    }
+
+    private CPDReport mockReportFor(Match match) {
+        List<Match> matches = singletonList(match);
+        CPDReport cpdReport = mock(CPDReport.class);
+        when(cpdReport.getMatches()).thenReturn(matches);
+        return cpdReport;
+    }
+
+    private void assertLanguage(CPDConfiguration configuration, String languageId) {
+        String fileThatShouldMatch;
+        String fileThatShouldNotMatch = "foo.txt";
+        switch (languageId) {
+            case "java":
+                fileThatShouldMatch = "foo.java";
+                break;
+            case "kotlin":
+                fileThatShouldMatch = "foo.kt";
+                break;
+            case "cpp":
+                fileThatShouldMatch = "foo.cpp";
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported language: " + languageId);
+        }
+        LanguageVersionDiscoverer languageVersionDiscoverer = configuration.getLanguageVersionDiscoverer();
+        assertThat(languageVersionDiscoverer.getLanguagesForFile(fileThatShouldMatch)).isNotEmpty();
+        assertThat(languageVersionDiscoverer.getLanguagesForFile(fileThatShouldNotMatch)).isEmpty();
     }
 }

--- a/src/test/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdReporterTest.java
+++ b/src/test/java/de/aaschmid/gradle/plugins/cpd/internal/worker/CpdReporterTest.java
@@ -1,21 +1,18 @@
 package de.aaschmid.gradle.plugins.cpd.internal.worker;
 
-import java.io.File;
-import java.io.IOException;
-import java.nio.file.Path;
-
 import de.aaschmid.gradle.plugins.cpd.internal.worker.CpdWorkParameters.Report;
 import de.aaschmid.gradle.plugins.cpd.test.TestTag;
+import net.sourceforge.pmd.cpd.CPDReport;
+import net.sourceforge.pmd.cpd.CPDReportRenderer;
 import net.sourceforge.pmd.cpd.CSVRenderer;
 import net.sourceforge.pmd.cpd.Mark;
 import net.sourceforge.pmd.cpd.Match;
 import net.sourceforge.pmd.cpd.SimpleRenderer;
-import net.sourceforge.pmd.cpd.SourceCode;
-import net.sourceforge.pmd.cpd.SourceCode.StringCodeLoader;
-import net.sourceforge.pmd.cpd.TokenEntry;
 import net.sourceforge.pmd.cpd.VSRenderer;
 import net.sourceforge.pmd.cpd.XMLRenderer;
-import net.sourceforge.pmd.cpd.renderer.CPDRenderer;
+import net.sourceforge.pmd.lang.document.Chars;
+import net.sourceforge.pmd.lang.document.FileLocation;
+import org.assertj.core.api.SoftAssertions;
 import org.gradle.api.GradleException;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
@@ -24,8 +21,11 @@ import org.junit.jupiter.api.io.TempDir;
 import org.mockito.InjectMocks;
 import org.mockito.junit.jupiter.MockitoExtension;
 
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+
 import static java.util.Arrays.asList;
-import static java.util.Collections.emptyList;
 import static java.util.Collections.singletonList;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
@@ -49,14 +49,14 @@ class CpdReporterTest {
         Report report = mock(Report.class);
         when(report.getDestination()).thenReturn(tempDir.resolve("report.file").toFile());
 
-        CPDRenderer cpdRenderer = mock(CPDRenderer.class);
+        CPDReportRenderer cpdRenderer = mock(CPDReportRenderer.class);
         doThrow(new IOException("foo")).when(cpdRenderer).render(any(), any());
 
         CpdReporter underTestSpy = spy(underTest);
         doReturn(cpdRenderer).when(underTestSpy).createRendererFor(report);
 
         // Expect:
-        assertThatThrownBy(() -> underTestSpy.generate(singletonList(report), emptyList()))
+        assertThatThrownBy(() -> underTestSpy.generate(singletonList(report), mock(CPDReport.class)))
                 .isInstanceOf(GradleException.class)
                 .hasMessage("foo")
                 .hasCauseInstanceOf(IOException.class);
@@ -82,22 +82,29 @@ class CpdReporterTest {
         File xmlReportFile = tempDir.resolve("cpd.xml").toFile();
         Report.Xml xmlReport = new Report.Xml(xmlReportFile, "ISO-8859-15");
 
+        Mark mark = mockMarkWithLocation(1, 2);
 
-        Mark mark = new Mark(new TokenEntry("1", "Clazz1.java", 1));
-        mark.setLineCount(1);
-        mark.setSourceCode(new SourceCode(new StringCodeLoader("String str = \"I am a duplicate\";")));
+        Match match = mock(Match.class);
+        when(match.getLineCount()).thenReturn(1);
+        when(match.getTokenCount()).thenReturn(5);
+        when(match.iterator()).thenAnswer(a -> asList(mark, mark).iterator());
 
-        Match match = new Match(5, mark, mark);
+        CPDReport cpdReport = mock(CPDReport.class);
+        when(cpdReport.getDisplayName(any())).thenReturn("Clazz1.java");
+        when(cpdReport.getMatches()).thenReturn(asList(match, match));
+        when(cpdReport.getSourceCodeSlice(any())).thenReturn(Chars.EMPTY);
 
         // When:
-        underTest.generate(asList(csvReport, csvReportWithoutLines, textReport, vsReport, xmlReport), asList(match, match));
+        underTest.generate(asList(csvReport, csvReportWithoutLines, textReport, vsReport, xmlReport), cpdReport);
 
         // Then:
-        assertThat(contentOf(csvReportFile)).startsWith("lines,tokens,occurrences\n");
-        assertThat(contentOf(csvReportFileWithoutLines)).startsWith("tokens;occurrences\n");
-        assertThat(contentOf(textReportFile)).startsWith("Found a 1 line (5 tokens) duplication in the following files: \n");
-        assertThat(contentOf(vsReportFile)).startsWith("Clazz1.java(1): Between lines 1 and 2\n");
-        assertThat(contentOf(xmlReportFile)).startsWith("<?xml version=\"1.0\" encoding=\"ISO-8859-15\"?>\n");
+        SoftAssertions.assertSoftly(softly -> {
+            softly.assertThat(contentOf(csvReportFile)).startsWith("lines,tokens,occurrences\n");
+            softly.assertThat(contentOf(csvReportFileWithoutLines)).startsWith("tokens;occurrences\n");
+            softly.assertThat(contentOf(textReportFile)).startsWith("Found a 1 line (5 tokens) duplication in the following files: \n");
+            softly.assertThat(contentOf(vsReportFile)).startsWith("Clazz1.java(1): Between lines 1 and 2\n");
+            softly.assertThat(contentOf(xmlReportFile)).startsWith("<?xml version=\"1.0\" encoding=\"ISO-8859-15\"?>\n");
+        });
     }
 
     @Test
@@ -106,7 +113,7 @@ class CpdReporterTest {
         Report.Csv report = new Report.Csv(new File("cpd.csv"), ';', true);
 
         // When:
-        CPDRenderer result = underTest.createRendererFor(report);
+        CPDReportRenderer result = underTest.createRendererFor(report);
 
         // Then:
         assertThat(result)
@@ -121,7 +128,7 @@ class CpdReporterTest {
         Report.Text report = new Report.Text(new File("cpd.txt"), "---", true);
 
         // When:
-        CPDRenderer result = underTest.createRendererFor(report);
+        CPDReportRenderer result = underTest.createRendererFor(report);
 
         // Then:
         assertThat(result)
@@ -136,7 +143,7 @@ class CpdReporterTest {
         Report.Vs report = new Report.Vs(new File("cpd.vs"));
 
         // When:
-        CPDRenderer result = underTest.createRendererFor(report);
+        CPDReportRenderer result = underTest.createRendererFor(report);
 
         // Then:
         assertThat(result).isInstanceOf(VSRenderer.class);
@@ -148,7 +155,7 @@ class CpdReporterTest {
         Report.Xml report = new Report.Xml(new File("cpd.xml"), "ISO-8859-1");
 
         // When:
-        CPDRenderer result = underTest.createRendererFor(report);
+        CPDReportRenderer result = underTest.createRendererFor(report);
 
         // Then:
         assertThat(result)
@@ -166,5 +173,15 @@ class CpdReporterTest {
         assertThatThrownBy(() -> underTest.createRendererFor(report))
                 .isInstanceOf(GradleException.class)
                 .hasMessage("Cannot create reports for unsupported type '" + report.getClass() + "'.");
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private Mark mockMarkWithLocation(int startLine, int endLine) {
+        Mark mark = mock(Mark.class);
+        FileLocation location = mock(FileLocation.class);
+        when(location.getStartLine()).thenReturn(startLine);
+        when(location.getEndLine()).thenReturn(endLine);
+        when(mark.getLocation()).thenReturn(location);
+        return mark;
     }
 }

--- a/src/test/resources/test-data/java/de/aaschmid/annotation/Employee.java
+++ b/src/test/resources/test-data/java/de/aaschmid/annotation/Employee.java
@@ -4,7 +4,7 @@ import javax.persistence.*;
 
 
 @Entity
-@Table(uniqueConstraints = {@UniqueConstraint(columnNames = { "name" })}
+@Table(uniqueConstraints = {@UniqueConstraint(columnNames = { "name" })},
         indexes= {
         @Index(name = "idx_name", columnNames = { "name" }),
         @Index(name = "idx_dayOfBirth", columnNames = { "dayOfBirth" }),

--- a/src/test/resources/test-data/java/de/aaschmid/annotation/Person.java
+++ b/src/test/resources/test-data/java/de/aaschmid/annotation/Person.java
@@ -4,7 +4,7 @@ import javax.persistence.*;
 
 
 @Entity
-@Table(uniqueConstraints = {@UniqueConstraint(columnNames = { "name" })}
+@Table(uniqueConstraints = {@UniqueConstraint(columnNames = { "name" })},
         indexes= {
         @Index(name = "idx_name", columnNames = { "name" }),
         @Index(name = "idx_dayOfBirth", columnNames = { "dayOfBirth" }),

--- a/src/test/resources/test-data/java/de/aaschmid/clazz/Clazz.java
+++ b/src/test/resources/test-data/java/de/aaschmid/clazz/Clazz.java
@@ -1,5 +1,5 @@
 package de.aaschmid.clazz;
 
 public interface Clazz {
-    public abstract isClazz(Clazz clazz);
+    boolean isClazz(Clazz clazz);
 }

